Libraries/LibPCIDB/Database.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <AK/HashMap.h>
+#include <AK/MappedFile.h>
+#include <AK/NonnullRefPtr.h>
+#include <AK/RefCounted.h>
+#include <AK/RefPtr.h>
+#include <AK/StringView.h>
+
+namespace PCIDB {
+
+struct Subsystem : public RefCounted<Subsystem> {
+    u16 vendor_id;
+    u16 device_id;
+    StringView name;
+};
+
+struct Device : public RefCounted<Device> {
+    u16 id;
+    StringView name;
+    HashMap<int, NonnullRefPtr<Subsystem>> subsystems;
+};
+
+struct Vendor : public RefCounted<Vendor> {
+    u16 id;
+    StringView name;
+    HashMap<int, NonnullRefPtr<Device>> devices;
+};
+
+struct ProgrammingInterface : public RefCounted<ProgrammingInterface> {
+    u8 id { 0 };
+    StringView name {};
+};
+
+struct Subclass : public RefCounted<Subclass> {
+    u8 id { 0 };
+    StringView name {};
+    HashMap<int, NonnullRefPtr<ProgrammingInterface>> programming_interfaces;
+};
+
+struct Class : public RefCounted<Class> {
+    u8 id { 0 };
+    StringView name {};
+    HashMap<int, NonnullRefPtr<Subclass>> subclasses;
+};
+
+class Database : public RefCounted<Database> {
+public:
+    static RefPtr<Database> open(const StringView& file_name = "/res/pci.ids") { return adopt(*new Database(file_name)); }
```

Comment:
The `RefPtr` return type implies that this could fail and return `nullptr` but it never does.
Perhaps this should be a helper with some more logic -- if it fails to open the provided path, return `nullptr`, otherwise return a new `Database`.